plugins/module_utils/common.py

```diff
@@ -432,19 +441,49 @@ def _custom_condition(resource):
         def _resource_absent(resource):
             return not resource
 
+        # wait_for requires jsonpath-rw library
+        jsonpath_expr = None
+        if wait_for is not None:
+            if not HAS_JSONPATH_RW:
+                if hasattr(self, 'fail_json'):
+                    self.fail_json(msg=missing_required_lib('jsonpath_rw'), error=to_native(jsonpath_import_exception))
+                raise WaitException("wait_for option requires 'jsonpath_rw' library")
+            try:
+                wait_expr = wait_for
+                if wait_for.startswith("."):
+                    wait_expr = "$" + wait_for
+                jsonpath_expr = jsonpath_rw.parse(wait_expr)
+            except Exception as parse_err:
+                if hasattr(self, 'fail_json'):
+                    self.fail_json(msg="Failed to parse wait_for attribute {0}".format(wait_for), error=to_native(parse_err))
+                raise WaitException("Failed to parse wait_for attribute {0} error is {1}".format(wait_for, to_native(parse_err)))
+
+        def _wait_for_property(resource):
+            try:
+                return all([match.value for match in jsonpath_expr.find(resource)])
+            except Exception as e:
+                return False
+
         waiter = dict(
             Deployment=_deployment_ready,
             DaemonSet=_daemonset_ready,
             Pod=_pod_ready
         )
         kind = definition['kind']
-        if state == 'present' and not condition:
-            predicate = waiter.get(kind, lambda x: x)
-        elif state == 'present' and condition:
-            predicate = _custom_condition
+        predicates = []
+        if state == 'present':
+            if condition is None and wait_for is None:
+                predicates.append(waiter.get(kind, lambda x: x))
+            else:
+                if condition:
+                    # add waiter on custom condition
+                    predicates.append(_custom_condition)
+                if wait_for:
+                    # json path predicate
+                    predicates.append(_wait_for_property)
         else:
-            predicate = _resource_absent
-        return self._wait_for(resource, definition['metadata']['name'], definition['metadata'].get('namespace'), predicate, sleep, timeout, state)
+            predicates.append(_resource_absent)
```

Comment:
I think we'd want to make sure that resource_absent is the only predicate in this case, the resource being waited on should be deleted so any other predicate in addition to that one will probably just make it wait until it times out